File: Android/tuilivekit/component/audioeffect/src/main/java/com/trtc/uikit/component/audioeffect/view/ReverbAdapter.java
package com.trtc.uikit.component.audioeffect.view;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.tencent.liteav.audio.TXAudioEffectManager.TXVoiceReverbType;
import com.trtc.uikit.component.audioeffect.R;
import com.trtc.uikit.component.audioeffect.service.AudioEffectService;

import java.util.ArrayList;
import java.util.List;

public class ReverbAdapter extends RecyclerView.Adapter<ReverbAdapter.ViewHolder> {

    private final Context          mContext;
    private       List<ReverbItem>   mData;
    private final AudioEffectService mAudioEffectService;
    private       int                mSelectedPosition;

    public ReverbAdapter(Context context, AudioEffectService audioEffectService) {
        mContext = context;
        mAudioEffectService = audioEffectService;
        initData();
    }

    private void initData() {
        mData = new ArrayList<>();
        mData.add(new ReverbItem(mContext.getString(R.string.livekit_reverb_none),
                R.drawable.livekit_select_none, TXVoiceReverbType.TXLiveVoiceReverbType_0));
        mData.add(new ReverbItem(mContext.getString(R.string.livekit_reverb_karaoke),
                R.drawable.livekit_reverb_ktv, TXVoiceReverbType.TXLiveVoiceReverbType_1));
        mData.add(new ReverbItem(mContext.getString(R.string.livekit_reverb_metallic_sound),
                R.drawable.livekit_reverb_metallic_sound, TXVoiceReverbType.TXLiveVoiceReverbType_6));
        mData.add(new ReverbItem(mContext.getString(R.string.livekit_reverb_low),
                R.drawable.livekit_reverb_low, TXVoiceReverbType.TXLiveVoiceReverbType_4));
        mData.add(new ReverbItem(mContext.getString(R.string.livekit_reverb_loud_and_loud),
                R.drawable.livekit_reverb_loud_and_loud, TXVoiceReverbType.TXLiveVoiceReverbType_5));
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View itemView = LayoutInflater.from(parent.getContext()).inflate(R.layout.livekit_anchor_settings_panel_item,
                parent, false);
        return new ViewHolder(itemView);
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        final ReverbItem reverbItem = mData.get(position);
        holder.textTitle.setText(reverbItem.title);
        holder.imageIcon.setImageResource(reverbItem.icon);
        if (reverbItem.type == mAudioEffectService.mAudioEffectState.reverbType.get()) {
            holder.imageIcon.setBackgroundResource(R.drawable.livekit_settings_item_select_background);
            mSelectedPosition = mData.indexOf(reverbItem);
        } else {
            holder.imageIcon.setBackgroundResource(R.drawable.livekit_settings_item_not_select_background);
        }
        holder.layoutRoot.setOnClickListener(view -> {
            int index = holder.getBindingAdapterPosition();
            if (index != RecyclerView.NO_POSITION) {
                mAudioEffectService.setVoiceReverbType(reverbItem.type);
                notifyItemChanged(index);
                notifyItemChanged(mSelectedPosition);
            }
        });
    }

    @Override
    public int getItemCount() {
        return mData.size();
    }

    public static class ViewHolder extends RecyclerView.ViewHolder {
        public LinearLayout layoutRoot;
        public TextView     textTitle;
        public ImageView    imageIcon;

        public ViewHolder(View itemView) {
            super(itemView);
            layoutRoot = itemView.findViewById(R.id.ll_root);
            textTitle = itemView.findViewById(R.id.tv_title);
            imageIcon = itemView.findViewById(R.id.iv_icon);
        }
    }

    public static class ReverbItem {
        public String            title;
        public int               icon;
        public TXVoiceReverbType type;

        public ReverbItem(String title, int icon, TXVoiceReverbType type) {
            this.title = title;
            this.icon = icon;
            this.type = type;
        }
    }
}

File: Android/tuilivekit/src/main/java/com/trtc/uikit/livekit/component/audioeffect/view/ReverbAdapter.kt
package com.trtc.uikit.livekit.component.audioeffect.view

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.trtc.uikit.livekit.R
import io.trtc.tuikit.atomicxcore.api.AudioEffectStore
import io.trtc.tuikit.atomicxcore.api.AudioReverbType

class ReverbAdapter(private val context: Context, private val audioEffectStore: AudioEffectStore) :
    RecyclerView.Adapter<ReverbAdapter.ViewHolder>() {

    private var data: MutableList<ReverbItem> = ArrayList()
    private var selectedPosition: Int = 0

    init {
        initData()
    }

    private fun initData() {
        data.add(
            ReverbItem(
                context.getString(R.string.common_reverb_none),
                R.drawable.audio_effect_select_none, AudioReverbType.NONE
            )
        )
        data.add(
            ReverbItem(
                context.getString(R.string.common_reverb_karaoke),
                R.drawable.audio_effect_reverb_ktv, AudioReverbType.KTV
            )
        )
        data.add(
            ReverbItem(
                context.getString(R.string.common_reverb_metallic_sound),
                R.drawable.audio_effect_reverb_metallic_sound,
                AudioReverbType.METALLIC
            )
        )
        data.add(
            ReverbItem(
                context.getString(R.string.common_reverb_low),
                R.drawable.audio_effect_reverb_low, AudioReverbType.DEEP
            )
        )
        data.add(
            ReverbItem(
                context.getString(R.string.common_reverb_loud_and_loud),
                R.drawable.audio_effect_reverb_loud_and_loud,
                AudioReverbType.LOUD
            )
        )
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val itemView = LayoutInflater.from(parent.context).inflate(
            R.layout.audio_effect_layout_reverb_item,
            parent, false
        )
        return ViewHolder(itemView)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val reverbItem = data[position]
        holder.textTitle.text = reverbItem.title
        holder.imageIcon.setImageResource(reverbItem.icon)
        if (reverbItem.type == audioEffectStore.audioEffectState.audioReverbType.value) {
            holder.imageIcon.setBackgroundResource(R.drawable.audio_effect_settings_item_select_background)
            selectedPosition = data.indexOf(reverbItem)
        } else {
            holder.imageIcon.setBackgroundResource(R.drawable.audio_effect_settings_item_not_select_background)
        }
        holder.layoutRoot.setOnClickListener { view ->
            val index = holder.bindingAdapterPosition
            if (index != RecyclerView.NO_POSITION) {
                audioEffectStore.setAudioReverbType(reverbItem.type)
                notifyItemChanged(index)
                notifyItemChanged(selectedPosition)
            }
        }
    }

    override fun getItemCount(): Int {
        return data.size
    }

    class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val layoutRoot: LinearLayout = itemView.findViewById(R.id.ll_root)
        val textTitle: TextView = itemView.findViewById(R.id.tv_title)
        val imageIcon: ImageView = itemView.findViewById(R.id.iv_icon)
    }

    class ReverbItem(
        val title: String,
        val icon: Int,
        val type: AudioReverbType
    )
}